File: leecodeStudy/checkPerfectNumber.py
# leecode 507

class CheckPerfectNumber_Solution:
    def checkPerfectNumber_V1(self, num : int) -> bool:
        if num == 1:
            return False
        upper_bound = num
        factor_sum = 1
        factor = 2
        while factor < upper_bound:
            if num % factor == 0:
                factor_sum += factor
                another_factor = num / factor
                if another_factor != factor:
                    factor_sum += another_factor
                    upper_bound = another_factor
                else:
                    break
            factor += 1        
        return num == factor_sum

    def checkPerfectNumber_V2(self, num : int) -> bool:
        return num in [6, 28, 496, 8128, 33550336]

File: leecodeStudy/checkPerfectNumber.cpp
#include <iostream>
bool checkPerfectNumber(int num){
    if (num == 1){
        return false;
    }
    int upper_bound = num / 2;
    int factor_sum = 1;
    int factor_maybe = 2;
    int another_factor = 0;
    while(factor_maybe < upper_bound){
        if (num % factor_maybe == 0){
            factor_sum += factor_maybe;
            another_factor = num / factor_maybe;
                if(another_factor != factor_maybe){
                    factor_sum += another_factor;
                    upper_bound = another_factor;
                }
                else{
                    break;
                }
        }
        factor_maybe += 1;
    }
    return num == factor_sum;
}

int main(){
    cout << checkPerfectNumber(28) << endl;
}